lib/raygun.dart

```diff
@@ -0,0 +1,226 @@
+
+    // Get messages and null check
+    G_FFIResult_FFIVec_Message messages = bindings.raygun_get_messages(
+        pRaygun, conversationID.toNativeUtf8().cast<Int8>());
+    if (messages.error.address.toString() != "0") {
+      throw WarpException(messages.error);
+    }
+
+    List<Message> msgs = [];
+    // Iterate over the messages
+    int length = messages.data.ref.len;
+    for (int i = 0; i < length; i++) {
+      Message message = Message();
+      Pointer<G_Message> pMsg = messages.data.ref.ptr.elementAt(i).value;
+
+      // Message ID
+      message.id = bindings.message_id(pMsg).value.toString();
+      // Conversation ID
+      message.conversationId =
+          bindings.message_conversation_id(pMsg).value.toString();
+      // Sender ID, DID only
+      Pointer<G_SenderId> pSenderId = bindings.message_sender_id(pMsg);
+      Pointer<G_DID> pSenderDid = bindings.sender_id_get_did_key(pSenderId);
+      message.senderId = DID(pSenderDid).toString();
+      // DateTime
+      // TODO: test required
+      // The value from Rust is integer (UTC from Chrono ).
+      // Possibly DateTime.fromMillisecondsSinceEpoch() can be used.
+      message.date = DateTime(bindings.message_date(pMsg).value);
+      // Pinned
+      message.pinned = bindings.message_pinned(pMsg);
+      // Reactions
+      List<Reaction> reactions = [];
+      Pointer<G_FFIVec_Reaction> pReactions = bindings.message_reactions(pMsg);
+      int reactionsLen = pReactions.ref.len;
+      for (int j = 0; j < reactionsLen; j++) {
+        Reaction reaction = Reaction();
+        Pointer<G_Reaction> pReaction = pReactions.ref.ptr.elementAt(j).value;
+        reaction.emoji = bindings.reaction_emoji(pReaction).toString();
+        Pointer<G_FFIVec_SenderId> pReactionSendersId =
+            bindings.reaction_users(pReaction);
+        int reactionSendersIdLen = pReactionSendersId.ref.len;
+        for (int k = 0; k < reactionSendersIdLen; k++) {
+          reaction.senderId
+              .add(pReactionSendersId.ref.ptr.elementAt(k).value.toString());
+        }
+      }
+      message.reactions = reactions;
+      // Replied
+      Pointer<Int8> pReplied = bindings.message_replied(pMsg);
+      if (pReplied.address.toString() != "0") {
+        message.replied = pReplied.toString();
+      }
+      // Message body
+      Pointer<G_FFIVec_String> pLines = bindings.message_lines(pMsg);
+      int lineLen = pLines.ref.len;
+      for (int j = 0; j < lineLen; j++) {
+        message.value
+            .add(pLines.ref.ptr.elementAt(j).cast<Utf8>().toDartString());
+      }
+
+      msgs.add(message);
+      bindings.message_free(pMsg);
+
+      // TODO: Metadata - Rust binding is not ready
+    }
+
+    return msgs;
+  }
+
+  // This method serves for sending and editing.
+  // If there is message sent, Warp will cover the editing part.
+  send(String conversationId, String messageId, Message messages) {
```

Comment:
I think I said this in the other review but the `messageId` here should be optional. When sending with one it would edit the message with the new message but with it being optional and passing nullptr in the ffi function, it would send a new message.

Reply:
1. Updated for the lines. I tried to allocate data in heap with calloc and arena (depends on purpose). Seems like the using keyword and arena is the way to go (see the links below..), but not sure my usage is correct yet. Need some tests for sure.
2. I thought you implied for the delete method. Now the same approach is applied for the send and reply methods (I believe the react/pin/embed require the message ID still in any situation. Let me know if not).

Links:
- https://github.com/dart-lang/sdk/issues/43403#issuecomment-691903433
- https://stackoverflow.com/questions/68018694/type-safe-memory-allocation-with-dart-ffi
- https://github.com/dart-lang/language/issues/1642#issuecomment-845204655